plonky2/src/hash/keccak.rs

```diff
@@ -18,21 +18,59 @@ pub const SPONGE_WIDTH: usize = SPONGE_RATE + SPONGE_CAPACITY;
 /// Keccak-256 pseudo-permutation (not necessarily one-to-one) used in the challenger.
 /// A state `input: [F; 12]` is sent to the field representation of `H(input) || H(H(input)) || H(H(H(input)))`
 /// where `H` is the Keccak-256 hash.
-pub struct KeccakPermutation;
-impl<F: RichField> PlonkyPermutation<F, KeccakHashConfig> for KeccakPermutation {
-    fn permute(input: [F; SPONGE_WIDTH]) -> [F; SPONGE_WIDTH]
-    where
-        [(); SPONGE_WIDTH]:,
-    {
-        let mut state = vec![0u8; SPONGE_WIDTH * size_of::<u64>()];
+#[derive(Copy, Clone, Default, Debug, PartialEq)]
+pub struct KeccakPermutation<F: RichField> {
+    state: [F; SPONGE_WIDTH],
+}
+
+impl<F: RichField> Eq for KeccakPermutation<F> {}
+
+impl<F: RichField> AsRef<[F]> for KeccakPermutation<F> {
+    fn as_ref(&self) -> &[F] {
+        &self.state
+    }
+}
+
+// TODO: Several implementations here are copied from
+// PoseidonPermutation; they should be refactored.
+impl<F: RichField> PlonkyPermutation<F> for KeccakPermutation<F> {
+    const RATE: usize = SPONGE_RATE;
+    const WIDTH: usize = SPONGE_WIDTH;
+
+    fn new<I: IntoIterator<Item = F>>(elts: I) -> Self {
+        let mut perm = Self {
+            state: [F::default(); SPONGE_WIDTH],
+        };
+        perm.set_from_iter(elts, 0);
+        perm
+    }
+
+    fn set_elt(&mut self, elt: F, idx: usize) {
+        self.state[idx] = elt;
+    }
+
+    fn set_from_slice(&mut self, elts: &[F], start_idx: usize) {
+        let begin = start_idx;
+        let end = start_idx + elts.len();
+        self.state[begin..end].copy_from_slice(elts);
+    }
+
+    fn set_from_iter<I: IntoIterator<Item = F>>(&mut self, elts: I, start_idx: usize) {
+        for (s, e) in self.state[start_idx..].iter_mut().zip(elts) {
+            *s = e;
+        }
+    }
```

Comment:
The doc in the Trait says that it would panic if `start_idx + elts.len() > WIDTH`, but I think it would only if `start_idx > WIDTH` no? i.e. regardless of the length of `elts`, as long as we can start iterating on self.state (even an empty iterator in the case `start_idx = WIDTH`), the zipped iterator may yield `None`, effectively doing nothing but the function wouldn't panic. Or am I missing something?

Reply:
Good catch.